File: 3D Touch/PreviewViewController.m
//
//  PreviewViewController.m
//  3D Touch
//
//  Created by Jay Versluis on 21/09/2015.
//  Copyright © 2015 Pinkstone Pictures LLC. All rights reserved.
//

#import "PreviewViewController.h"

@interface PreviewViewController () <UIPreviewActionItem>

@end

@implementation PreviewViewController

- (void)viewDidLoad {
    
    [super viewDidLoad];
    [self check3DTouch];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)dismissMe{
    
    // dismiss this view controller
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)check3DTouch {
    
    // if 3D Touch is not available, add a tap gesture to dismiss this view
    if (self.traitCollection.forceTouchCapability != UIForceTouchCapabilityAvailable) {
        
        UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc]initWithTarget:self action:@selector(dismissMe)];
        [self.view addGestureRecognizer:tap];
    }
}


#pragma mark - Preview Actions

- (NSArray<id<UIPreviewActionItem>> *)previewActionItems {
    
    [super previewActionItems];
    
    // setup a list of preview actions
    UIPreviewAction *action1 = [UIPreviewAction actionWithTitle:@"Action 1" style:UIPreviewActionStyleDefault handler:^(UIPreviewAction * _Nonnull action, UIViewController * _Nonnull previewViewController) {
        NSLog(@"Action 1 triggered");
    }];
    
    UIPreviewAction *action2 = [UIPreviewAction actionWithTitle:@"Destructive Action" style:UIPreviewActionStyleDestructive handler:^(UIPreviewAction * _Nonnull action, UIViewController * _Nonnull previewViewController) {
        NSLog(@"Destructive Action triggered");
    }];
    
    UIPreviewAction *action3 = [UIPreviewAction actionWithTitle:@"Selected Action" style:UIPreviewActionStyleSelected handler:^(UIPreviewAction * _Nonnull action, UIViewController * _Nonnull previewViewController) {
        NSLog(@"Selected Action triggered");
    }];
    
    // add them to an arrary
    NSArray *actions = @[action1, action2, action3];
    
    // and return them
    return actions;
}

@end

File: 3D Touch - Swift/3D Touch Swift/PreviewViewController.swift
//
//  PreviewViewController.swift
//  3D Touch Swift
//
//  Created by Jay Versluis on 04/12/2015.
//  Copyright © 2015 Pinkstone Pictures LLC. All rights reserved.
//

import UIKit

class PreviewViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        self.check3DTouch()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func dismissMe() {
        
        // dismiss this view controller
        self.dismiss(animated: true, completion: nil)
    }
    
    func check3DTouch() {
        
        // if 3D Touch is not available, add a tap gesture to dismiss this controller
        if self.traitCollection.forceTouchCapability != UIForceTouchCapability.available {
            let tap = UITapGestureRecognizer.init(target: self, action: #selector(PreviewViewController.dismissMe))
            self.view.addGestureRecognizer(tap)
        }
    }
    
    // MARK: Preview Actions
    
    override var previewActionItems : [UIPreviewActionItem] {
    
        // setup a list of preview actions
        let action1 = UIPreviewAction.init(title: "Action 1", style: UIPreviewActionStyle.default) { (UIPreviewAction, UIViewController) -> Void in
            NSLog("Action 1 was selected")
        }
        
        let action2 = UIPreviewAction.init(title: "Destructive Action", style: UIPreviewActionStyle.destructive) { (UIPreviewAction, UIViewController) -> Void in
            NSLog("Destructive Action was selected")
        }
        
        let action3 = UIPreviewAction.init(title: "Selected Action", style: UIPreviewActionStyle.selected) { (UIPreviewAction, UIViewController) -> Void in
            NSLog("Selected Action was selected")
        }
        
        // add them to an array
        let actions = NSArray.init(objects: action1, action2, action3)
        
        // add all actions to a group
        let group = UIPreviewActionGroup.init(title: "Action Group", style: UIPreviewActionStyle.default, actions: actions as! [UIPreviewAction])
        
        // add the group to an array (yes, this is indeed ridiculous)
        let groupedGroup = NSArray.init(object: group)
        
        // and return them (return the array of actions instead to see all items ungrouped)
        return groupedGroup as! [UIPreviewActionItem]

    }

}
